File: 01Ruby/W1D2_Ander/students_and_courses.rb
class Courses
  attr_reader :course_name, :department, :credits
  attr_accessor :students

  def initialize(course_name, department, credits)
    @course_name = course_name
    @department = department
    @credits = credits
    @students = []
  end

  def student
    names = []
    students.each { |student| names << student.name }
    names * ", "
  end

  def add_student(student)
    unless students.include?(student)
      students << student
      student.enroll(self)
    end
  end

end

class Student
  attr_accessor :courses
  attr_reader :first_name, :last_name

  def initialize(first_name, last_name)
    @first_name = first_name
    @last_name = last_name
    @courses = []
  end

  def name
    "#{first_name} #{last_name}"
  end

  def enroll(course)
    unless courses.include?(course)
      courses << course
      course.add_student(self)
    end
  end

  def course_load
    course_hash = Hash.new(0)
    courses.each do |course|
      # if course_hash.include?(course.department)
        course_hash[course.department] += course.credits
      # else
        # course_hash[course.department] = course.credits
      # end
    end

    course_hash
  end

end

File: 04JS/W5D4/students-and-courses.js
function Student(firstName, lastName){
  this.firstName = firstName;
  this.lastName = lastName;
  this.courses = [];
}

function Course(name, department, credits){
  this.name = name;
  this.department = department;
  this.credits = credits;
  this.students = [];
}

Student.prototype.name = function(){
  return this.firstName + " " + this.lastName;
};

Student.prototype.enroll = function(course){
  if (this.courses.indexOf(course) === -1) {
    this.courses.push(course);
  }
};

Course.prototype.addStudent = function(student){
  if (this.students.indexOf(student) === -1) {
    this.students.push(student);
    student.enroll(this);
  };
};

Student.prototype.courseLoad = function() {
  var studentLoad = {};
  this.courses.forEach (function(course) {
    if (studentLoad[course.department] === undefined ) {
      studentLoad[course.department] = course.credits;
    } else {
      studentLoad[course.department] += course.credits;
    }
  });
  return studentLoad;
};

var pat = new Student("Pat", "Lo");
var sherry = new Student("Sherry", "Zhou");
var algebra = new Course("Algebra", "Math", 4);
var biology = new Course("Biology", "Science", 4);
var calculus = new Course("Calculus", "Math", 5);

algebra.addStudent(pat);
algebra.addStudent(sherry);
biology.addStudent(pat);
calculus.addStudent(sherry);
